npm/design-system/src/components/FileExplorer/FileExplorer.tsx

```diff
@@ -0,0 +1,177 @@
+import React from 'react'
+import cs from 'classnames'
+import { FileNode, FolderNode, TreeNode } from './helpers/makeFileHierarchy'
+
+export interface FolderComponentProps {
+  item: FolderNode
+  depth: number
+  isOpen: boolean
+  onClick: () => void
+}
+
+export interface FileComponentProps {
+  item: FileNode
+  depth: number
+  onClick: (file: FileNode) => void
+}
+
+export interface FileExplorerProps extends React.HTMLAttributes<HTMLDivElement> {
+  files: TreeNode[]
+  fileComponent: React.FC<FileComponentProps>
+  folderComponent: React.FC<FolderComponentProps>
+  selectedFile?: string
+  onFileClick: (file: FileNode) => void
+
+  // Styles. They should be a *.module.scss.
+  // TODO: Can we type these? Do we want to couple to CSS modules?
+  cssModule?: {
+    nav: any
+    ul: any
+    li: any
+    a: any
+    isSelected: any
+  }
+}
+
+export interface FileTreeProps extends FileExplorerProps {
+  depth: number
+  openFolders: Record<string, boolean>
+  style?: React.CSSProperties
+  setSelectedFile: (absolute: string) => void
+}
+
+export const FileExplorer: React.FC<FileExplorerProps> = (props) => {
+  /**
+   * Whether a folder is open or not is a **UI** concern.
+   * From a file system point of view, there is no such concept as "open" or "closed",
+   * only from a user's point of view.
+   * For this reason we save the open state as part of the UI component. The easiest
+   * way to do this is a key/value pair, mapping the absolute path of a directory to a boolean
+   *
+   * {
+   *   'foo': true,
+   *   'foo/bar': true
+   *   'foo/bar/qux': false
+   * }
+   *
+   * Every directory is set to open by default. When you add a new directory
+   * or file via your file system (eg mkdir foo/bar && touch foo/bar/hello.js) it will be added
+   * without losing the current state of open/closed directories.
+   */
+  const [openFolders, setOpenFolders] = React.useState<Record<string, boolean>>({})
+
+  React.useEffect(() => {
+    function walk (nodes: TreeNode[]) {
+      for (const node of nodes) {
+        if (node.type === 'folder') {
+          if (!(node.absolute in openFolders)) {
```

Comment:
If you want this to not override previously collapsed folders (which is what this does), you should probably call it out in a comment.